Non-exclusive page write locks and support multiple open page cursors per thread

chrisvest opened the pull request:
- [x] Implement new lock primitive that support non-exclusive write locks
- [x] Replace StampedLock with new lock primitive in page cache
- [x] Make all existing tests pass
- [x] Add tests to verify that page write locks are non-exclusive
- [x] Verify all update code paths in `MetadataStore` are safe with non-exclusive write locks
- [x] Lift restriction on single open page cursor per thread, and add tests for it as well
- [x] Add randomised/fuzz test that uses multiple open page cursors per thread
- [x] Benchmark and profile the changes in performance that these changes make

Page write locks are now non-exclusive. This means that multiple page cursors can have a write lock on a page at the same time. This is fine because the records are protected by the entity locks; by our lock manager as part of transaction handling, basically.

Shut down is also a bit trickier now. We need to make sure that we stop all writes prior to unmapping paged files and shutting the page cache down, since we can otherwise loose writes that happen concurrently with shut down. Code inspection reveals that this is currently not a problem that we have.

After this is done and merged, we can look at the layers above the page cache, and remove the hoops they jump through to ensure that they only ever have a single open page cursor at a time.

Implemented in collaboration with @davidegrohmann

chrisvest committed:
First draft of the OptiLock, a replacement for StampedLock for when we want to allow more than one open page cursor per thread.

chrisvest committed:
Javadocs and stress test for the OptiLock

chrisvest committed:
Make MuninnPage extend OptiLock and make things *compile*

chrisvest committed:
Make the fast page cache tests pass

- OptiLock.writeLock has been changed to tryWriteLock and no longer blocks.
  This means OptiLock is now entirely non-blocking and allocation free.
- A bunch of data race bugs has been fixed. Particularly around page faulting
  and pinning. And eviction. And background flushing.
- Pages are now always exclusively locked when they are in the freelist. They
  are locked by eviction and unlocked by page faulting. This solves a race
  where a page pin, an eviction and a page fault happen concurrently and
  interleave in a particular way.
- RandomPageCacheTestHarness now also simulates entity locks, and ensures that
  records are locked when written to.
- Flushing a paged file will now spin on exclusively locked pages in its
  translation table. This way, the flush won't skip pages that are being looked
  at by the eviciton thread, or the background flush thread. This is important
  because the flushes might fail, in which case we don't want to miss dirty
  pages since this can lead to lost data. Especially if we are flushing the
  file because we want to unmap it.

The slow page cache tests are still failing, and we still haven't allowed
multiple open page cursors per thread.

chrisvest committed:
Make the page cache stress test pass as well. Now all tests pass after the introduction of OptiLock

chrisvest committed:
Delete the now unused StampedLock

chrisvest committed:
Remove Java7 support from UnsafeUtil

chrisvest committed:
Page cache javadoc updates since OptiLock

chrisvest committed:
Rename PF_SHARED_LOCK and PF_EXCLUSIVE_LOCK to PF_SHARED_READ_LOCK and PF_SHARED_WRITE_LOCK respectively.

chrisvest committed:
Protect writes to MetaDataStore with record locks

We do this now, because page write locks are no longer exclusive. This *might* not be necessary, but it's easier to reason about correctness when single- and multi-field MetaDataStore updates are atomic.

chrisvest committed:
Add page cache tests for the non-exclusive nature of shared write locks

chrisvest committed:
Remove the restriction that a thread cannot have more than one page cursor open at a time

This also basically disables the cursor pooling we had before. We'll have to benchmark and profile to see if that's a problem or not.
We could start implementing per-PagedFile cursor pools, and maybe simplify the cursor initialisation.

chrisvest committed:
Extract method to help inlining, maybe

chrisvest committed:
Add a randomised test that opens multiple page cursors

chrisvest committed:
Add a test and a fix for the live-lock situation where we want to evict a page, but they are all locked

chrisvest committed:
Make SingleFilePAgeSwapperTest pass when run from Intellij

chrisvest committed:
Add support for multiple open page cursors to the RandomPageCacheTestHarness

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OptiLock.java:
```diff
@@ -0,0 +1,229 @@
+ * The exclusive lock will also invalidate the optimistic read locks, but not the write locks. The exclusive lock is
+ * try-lock only, and will never block.
+ */
+public class OptiLock
```
Class name is a bit cryptic and `keep an air of mystery about it`. Even if you can read a java doc and get the general idea why it's called like that would propose to rename it to OptimisticWriteStampedLock or similar

chrisvest replied:
We could call it `SequenceLock` as a reference to the fundamental idea.

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OptiLock.java:
```diff
@@ -0,0 +1,229 @@
+ */
+public class OptiLock
+{
+    private static final long CNT_BITS = 17; // Bits for counting concurrent write-locks
```
would be nice to have a java doc why we need 17 bits

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OptiLock.java:
```diff
@@ -0,0 +1,229 @@
+{
+    private static final long CNT_BITS = 17; // Bits for counting concurrent write-locks
+
+    private static final long SEQ_BITS = 64 - 1 - CNT_BITS;
```
would prefer to see a constant for 64

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OptiLock.java:
```diff
@@ -0,0 +1,229 @@
+
+    /**
+     * Start an optimistic critical section, and return a stamp that can be used to validate if the read-lock was
+     * consistent. That is, if no write or exclusive lock was overlapping with the optimistic read-lock.
```
What if there was write or exclusive lock?

chrisvest replied:
Those bits are masked off and it won't validate. That's what this is about:

> That is, if no write or exclusive lock was overlapping with the optimistic read-lock.

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OptiLock.java:
```diff
@@ -0,0 +1,229 @@
+        long excl = s >>> CNT_BITS + SEQ_BITS;
+        long cnt = (s & CNT_MASK) >> SEQ_BITS;
+        long seq = s & SEQ_MASK;
+        StringBuilder sb = new StringBuilder( "OptiLock[E:" ).append( excl );
```
Would like to see describe state more descriptive, so it will not require to open source code to decrypt it.

MishaDemianenko commented on community/unsafe/src/main/java/org/neo4j/unsafe/impl/internal/dragons/UnsafeUtil.java:
```diff
@@ -262,35 +223,12 @@ public static long getFieldOffset( Class<?> type, String field )
     }
 
-    private static int getAndAddInt_java7( Object obj, long offset, int delta )
+    public static void loadFence()
```
a bit of java doc at least from http://openjdk.java.net/jeps/171 would be nice

MishaDemianenko commented on community/primitive-collections/src/main/java/org/neo4j/concurrent/BinaryLatch.java:
```diff
@@ -90,6 +90,18 @@ private void unparkSuccessor( Node waiters )
+     * @param blocker The blocker that is given as the reason for blocking the thread, accessible through
+     * {@link java.util.concurrent.locks.LockSupport#getBlocker(Thread)}.
+     */
+    public void await( Object blocker )
```
on a review time moment there is no usages of newly introduced method, that is imho redundant

chrisvest replied:
Yeah, it was introduced to support blocking methods in OptiLock, but then those blocking methods got removed and this guy remained. I'll remove it.

chrisvest committed:
Rename OptiLock to SequenceLock and other small cleanups

chrisvest committed:
Add a "freeze" lock mode to the SequenceLock

And make the background and foreground flushing use it.
This new lock mode prevents write and exclusive locks, but does not invalidate read locks.
This reduces the negative impact the background flush task has on the performance of reads.

chrisvest committed:
Various optimisations

- Reintroduce cursor pooling, and trade object allocation for ThreadLocal
  lookups. Turns out this helps a lot.
- Reorganise the conditional branching in `MuninnPageCursor.pin()` to be more
  straight-line and have less branching in general.
- Also extract uncommon paths out of `pin()` into their own methods.
- A few unnecessary volatile annotations have been removed.
- Removed binary lock state representation from SequenceLock.
- Add a freeze lock mode to SequenceLock that fail write and exclusive locks,
  but not read locks. Useful for flushing.

This seems to put performance about on par with our previous StampedLock based implementation.

chrisvest committed:
Fix a test configuration for one of the page cache tests

The `PageCacheSlowTest.mustNotLoseUpdatesWhenOpeningMultiplePageCursorsPerThread()` test was flaky and would randomly fail.
It had been carelessly configured such that the number of threads accessing the page cache, multiplied by the maximum number of page cursors they could open, was exceding the number of cache pages.
This meant that it was possible for the test to run into live locks, where a page fault would fruitlessly look for a page to cooperatively evict, only to discover that all pages in the cache were already locked for other purposes.
The test is now tuned such that this cannot occur - not even if the test repitition count is dramatically increased; not even if the live-lock threshold is dramatically decreased.

MishaDemianenko commented:
@chrisvest there is a failed test case

chrisvest commented:
@MishaDemianenko Just pushed a fix.
